File: 2023/day01/1.py
num_map = {
    "one": "1", "two": "2", "three": "3",
    "four": "4", "five": "5", "six": "6",
    "seven": "7", "eight": "8", "nine": "9"
}

def parse_input(input):
    return input.readlines()

def solve1(input):
    total = 0
    first, last = 0, 0
    for line in input:
        filtered = list(filter(str.isdigit, line))
        if len(filtered) == 1:
            first, last = filtered[0], filtered[0]
        else:
            first, *_,  last = filtered
        total += int(first + last)
    return total
    

def solve2(input):
    total = 0
    for line in input:
        digits_found = []
        for index in range(len(line)):
            if line[index].isdigit():
                digits_found.append(line[index])
            for key, value in num_map.items():
                if index + len(key) > len(line):
                    continue
                if line[index:index+len(key)] == key:
                    digits_found.append(value)
        if len(digits_found) == 1:
            first, last = digits_found[0], digits_found[0]
        else:
            first, *_, last = digits_found
        total += int(first + last)
    return total

print("Part 1:", solve1(parse_input(open("1.in"))))
print("Part 2:", solve2(parse_input(open("1.in"))))

File: 2023/day01/1.exs
nums = %{
  one: "1",
  two: "2",
  three: "3",
  four: "4",
  five: "5",
  six: "6",
  seven: "7",
  eight: "8",
  nine: "9"
}

replacements = %{
  one: "onee",
  two: "twoo",
  five: "fivee",
  seven: "sevenn",
  eight: "eightt",
  nine: "ninee"
}

part1 =
  File.stream!("1.in")
  |> Enum.map(&String.trim/1)
  |> Enum.map(&String.replace(&1, ~r/[^\d]/, ""))
  |> Enum.map(fn val -> if val != "", do: "#{String.first(val)}#{String.last(val)}", else: "0" end)
  |> Enum.map(&String.to_integer/1)
  |> Enum.sum()

part2 =
  File.stream!("1.in")
  |> Enum.map(&String.trim/1)
  |> Enum.map(fn line ->
    Map.keys(replacements)
    |> Enum.reduce(
      line,
      fn key, acc ->
        str_key = Atom.to_string(key)

        if String.contains?(line, str_key) do
          String.replace(acc, str_key, Map.get(replacements, key))
        else
          acc
        end
      end
    )
  end)
  |> Enum.map(fn line ->
    Map.keys(nums)
    |> Enum.reduce(
      line,
      fn key, acc ->
        String.replace(acc, Atom.to_string(key), Map.get(nums, key))
      end
    )
  end)
  |> Enum.map(&String.replace(&1, ~r/[^\d]/, ""))
  |> Enum.map(fn val -> if val != "", do: "#{String.first(val)}#{String.last(val)}", else: "0" end)
  |> Enum.map(&String.to_integer/1)
  |> Enum.sum()

IO.puts("Part 1: #{part1}")
IO.puts("Part 2: #{part2}")
